Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt

```diff
@@ -0,0 +1,49 @@
+package de.rki.coronawarnapp.util.viewmodel
+
+import android.os.Bundle
+import androidx.lifecycle.AbstractSavedStateViewModelFactory
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.ViewModel
+import androidx.savedstate.SavedStateRegistryOwner
+import com.squareup.inject.assisted.Assisted
+import com.squareup.inject.assisted.AssistedInject
+
+class CWAViewModelSource @AssistedInject constructor(
+    private val creators: @JvmSuppressWildcards Map<Class<out CWAViewModel>, CWAViewModelFactory<out CWAViewModel>>,
+    @Assisted savedStateOwner: SavedStateRegistryOwner,
+    @Assisted defaultSavedState: Bundle?,
+    @Assisted private val assistAction: ((CWAViewModelFactory<out CWAViewModel>, SavedStateHandle) -> CWAViewModel)?
+) : AbstractSavedStateViewModelFactory(savedStateOwner, defaultSavedState) {
+
+    override fun <T : ViewModel?> create(
+        key: String,
+        modelClass: Class<T>,
+        handle: SavedStateHandle
+    ): T {
+        val factory = creators.entries.find { modelClass.isAssignableFrom(it.key) }?.value
+        if (factory == null) throw IllegalStateException("Unknown ViewModel factory: $modelClass")
+
+        @Suppress("UNCHECKED_CAST")
+        var vm: T? = assistAction?.invoke(factory, handle) as? T
+
+        @Suppress("UNCHECKED_CAST")
+        if (vm == null) {
+            vm = when (factory) {
+                is SavedStateCWAViewModelFactory<*> -> factory.create(handle) as? T
+                is SimpleCWAViewModelFactory<*> -> factory.create() as? T
```

Comment:
is there not an option to just use the handle as optional parameter and then use a generic factory?

Reply:
These default factories are used when a viewmodel doesn't need extra assisted arguments, e.g. so the fragment can have

```kotlin
class Fragment .... {
    @AssistedInject.Factory
    interface Factory : SavedStateVDCFactory<GeneratorsFragmentVDC>

}
```

instead of 

```kotlin
class StrawberryViewModel .... {
    @AssistedInject.Factory
    interface Factory : SavedStateVDCFactory<GeneratorsFragmentVDC> {
         fun create(handle: SavedStateHandle): StrawberryViewModel
    }

}
```

Default parameters wouldn't work as it would then still require any user of `SimpleCWAViewModelFactory` to explicitly implement that parameter, i.e.

```kotlin
class ViewModel(private val iDontNeedTheSavedState: SavedStateHandle?)
```